src/signature/mod.rs

```diff
@@ -8,124 +8,41 @@ pub mod mechanism;
-    /// the given image. The `PolicyRequirements` is defined in
-    /// [`policy_path`] and may include signature verification.
-    #[cfg(all(feature = "getresource", feature = "signature"))]
-    pub async fn allows_image(
-        image_reference: &str,
-        image_digest: &str,
-        secure_channel: Arc<Mutex<SecureChannel>>,
-        auth: &RegistryAuth,
-        paths: &Paths,
-    ) -> Result<()> {
-        // if Policy config file does not exist, get if from KBS.
-        if !std::path::Path::new(&paths.policy_path).exists() {
-            secure_channel
-                .lock()
-                .await
-                .get_resource(
-                    "Policy",
-                    std::collections::HashMap::new(),
-                    &paths.policy_path,
-                )
-                .await?;
-        }
-
-        let reference = oci_distribution::Reference::try_from(image_reference)?;
-        let mut image = Image::default_with_reference(reference);
-        image.set_manifest_digest(image_digest)?;
-
-        // Read the set of signature schemes that need to be verified
-        // of the image from the policy configuration.
-        let mut policy = Policy::from_file(&paths.policy_path).await?;
-        let schemes = policy.signature_schemes(&image);
-
-        // Get the necessary resources from KBS if needed.
-        for scheme in schemes {
-            scheme.init(paths).await?;
-            let resource_manifest = scheme.resource_manifest();
-            for (resource_name, path) in resource_manifest {
-                secure_channel
-                    .lock()
-                    .await
-                    .get_resource(resource_name, std::collections::HashMap::new(), path)
-                    .await?;
-            }
-        }
-
-        policy
-            .is_image_allowed(image, auth)
-            .await
-            .map_err(|e| anyhow::anyhow!("Validate image failed: {:?}", e))
+use crate::{config::Paths, signature::policy::Policy};
+use std::convert::TryFrom;
+
+use anyhow::Result;
+use oci_distribution::secrets::RegistryAuth;
+
+/// `allows_image` will check all the `PolicyRequirements` suitable for
+/// the given image. The `PolicyRequirements` is defined in
+/// [`policy_path`] and may include signature verification.
+#[cfg(feature = "signature")]
+pub async fn allows_image(
+    image_reference: &str,
+    image_digest: &str,
+    auth: &RegistryAuth,
+    file_paths: &Paths,
+) -> Result<()> {
+    use crate::{resource, signature::image::Image};
+
+    let reference = oci_distribution::Reference::try_from(image_reference)?;
+    let mut image = Image::default_with_reference(reference);
+    image.set_manifest_digest(image_digest)?;
+
+    // Read the set of signature schemes that need to be verified
+    // of the image from the policy configuration.
+    let policy_json_string = resource::get_resource(&file_paths.policy_path).await?;
```

Comment:
**BUG**: Here `&file_paths.policy_path` is `/run/image-security/security_policy.json`, but `get_resource` parameter needs KBS resource URI like `kbs://...`, right?

Reply:
Yes, thanks! fixed!